src/compiler/checker.ts

```diff
@@ -30855,7 +30861,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const jsxOpeningLikeNode = node ;
             const sig = getResolvedSignature(jsxOpeningLikeNode);
             checkDeprecatedSignature(sig, node);
-            checkJsxReturnAssignableToAppropriateBound(getJsxReferenceKind(jsxOpeningLikeNode), getReturnTypeOfSignature(sig), jsxOpeningLikeNode);
+
+            const elementTypeConstraint = getJsxElementTypeTypeAt(jsxOpeningLikeNode);
+            if (elementTypeConstraint !== undefined) {
+                const tagType = isJsxIntrinsicIdentifier(jsxOpeningLikeNode.tagName)
+                    ? // TODO: Should this be a literal that library authors could potentially check against?
+                        stringType
+                    : getApparentType(checkExpression(jsxOpeningLikeNode.tagName));

```

Comment:
You're using `getApparentType` here, but can you provide an example of what doesn't work without it?

For example, if you wrote

```ts
function f1<T extends (props: {}) => React.ReactElement>(Component: T) {
    return <Component />
}
```

I would think that this works fine without `getApparentType`, right?

Regardless if it does, can you add it to the test suite?

Reply:
Only used `getApparentType` because that was used in other places type-checking JSX. I removed it for now, since tests work without `getApparentType`.

The code you mentioned is [already rejected in 5.1. and stable React declarations](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdBArgHb7AQtwECMAPABU4WAB4wsLACboAFGGJhUALjgBvAL4BKOAF4AfJiowAdNjwwAogBssICTD7IWATz16ZAYRKQW9lQO1VOjg4ShgmKC4+L3BOezgAej16dSA).

I added the test regardless to verify we still have the same behavior.